AsyncDisplayKit/ASDisplayNode.mm

```diff
@@ -703,6 +704,89 @@ - (instancetype)styledWithBlock:(void (^)(ASLayoutElementStyle *style))styleBloc
 
 #pragma mark - Layout
 
+- (void)invalidateSize
+{
+  ASDisplayNodeAssertThreadAffinity(self);
+  
+  __instanceLock__.lock();
+
+  // Mark the node for layout in the next layout pass
+  [self invalidateCalculatedLayout];
+  
+  if (_supernode) {
+    ASDisplayNode *supernode = _supernode;
+    __instanceLock__.unlock();
+    // Cause supernode's layout to be invalidated
+    // We need to release the lock to prevent a deadlock
+    [supernode invalidateSize];
+    return;
+  }
+  
+  // Calculate a new pending layout. It will be applied in the next layout pass
+  ASLayout *layout = [self layoutThatFits:_calculatedDisplayNodeLayout->constrainedSize];
+  if (CGSizeEqualToSize(self.bounds.size, layout.size) == NO) {
+    // If the size of the layout changes inform the node of this
+    [self displayNodeDidInvalidateSizeOldSize:self.bounds.size];
+  }
+  
+  
+  __instanceLock__.unlock();
+}
+
+- (void)displayNodeDidInvalidateSizeOldSize:(CGSize)size
+{
+  // The default implementation of display node changes the size of itself to the new size
+  CGRect oldBounds = self.bounds;
```

Comment:
Should clarify the name of this method that it is only ever called for the root node. It would be a good idea to consider if we can make it a class method and passing all of the data necessary, this would keep it as an internal feature only for now but would be one step towards  and API to handle these events